### ISSUE TITLE
`torch.compile` silently produces all-zero `x.grad` for `diagonal_scatter(...).sum().backward()`

### ISSUE THREAD

**Issue by wuyii8941** [pytorch/pytorch#182012]
## 🐛 Describe the bug

`torch.compile(backend='inductor')` silently produces `x.grad = 0` for
`torch.diagonal_scatter(x, src, offset).sum().backward()`, while eager mode
correctly produces `x.grad = 1` at all non-diagonal positions (and `0` on the
diagonal, since those positions were overwritten by `src`).

The forward output is correct; only the backward gradient for `x` is wrong.
`src.grad` is computed correctly. `aot_eager` also produces the correct gradient,
confirming the bug is in Inductor's code generation, not in AOTAutograd or the
decomposition.

## Minimal reproducer

```python
import torch

x = torch.randn(8, 8, device='cuda', requires_grad=True)
src = torch.randn(8, device='cuda', requires_grad=True)

def fn(x, src):
    return torch.diagonal_scatter(x, src, 0).sum()

# Eager: correct
x1 = x.clone().detach().requires_grad_(True)
s1 = src.clone().detach().requires_grad_(True)
fn(x1, s1).backward()
print(f"eager x.grad:\n{x1.grad}")
# tensor([[0., 1., 1., 1., 1., 1., 1., 1.],
#         [1., 0., 1., 1., 1., 1., 1., 1.],
#         ...])  # 1 everywhere except diagonal

# Compiled: wrong
torch._dynamo.reset()
x2 = x.clone().detach().requires_grad_(True)
s2 = src.clone().detach().requires_grad_(True)
torch.compile(fn)(x2, s2).backward()
print(f"compiled x.grad:\n{x2.grad}")
# tensor([[0., 0., 0., 0., 0., 0., 0., 0.],
#         [0., 0., 0., 0., 0., 0., 0., 0.],
#         ...])  # ALL ZEROS — incorrect
```

## Root cause analysis

The backward FX graph for `diagonal_scatter(x, src).sum()` is:

```
expand(tangent_scalar, [N, N])  →  clone  →  diagonal_scatter(clone, zeros)  →  return
```

Inductor lowers `clone(expand(scalar, [N, N]))` into a Triton kernel that
writes a single element to `buf0`, then returns
`reinterpret_tensor(buf0, (N, N), (0, 0), 0)` — a stride-(0, 0) view where
all N² positions alias one memory cell. The subsequent `diagonal_scatter`
kernel writes `0.0` to that single cell (to zero the diagonal), so the entire
gradient reads back as 0.

Correct behavior: `clone` must materialize the expanded tensor into a
contiguous buffer so that `diagonal_scatter` can selectively zero only the
diagonal positions.

This is **not** the same root cause as #180164 / #180771 (slice_scatter
backward), which was in AOTAutograd's `meta_slice_scatter` preserving zero
strides. Here `aot_eager` is correct; only Inductor's codegen is wrong.

## Affected configurations

- **Shapes**: all tested sizes (4×4 through 256×256), square and non-square, all diagonal offsets
- **Dtypes**: float32 and float64
- **Devices**: CUDA (tested on Tesla T4)
- **Trigger pattern**: `diagonal_scatter(x, src, offset)` followed directly by `.sum()` (any dim variant)
- Adding a pointwise op between `diagonal_scatter` and `sum` (e.g. `.relu()`, `.mul(2)`) avoids the bug by preventing the problematic fusion

## Characterization

| Pattern | x.grad |
|---------|--------|
| `diagonal_scatter(x, src, 0).sum().backward()` | **all zeros (BUG)** |
| `diagonal_scatter(x, src, 0).relu().sum().backward()` | correct |
| `diagonal_scatter(x, src, 0).mul(2).sum().backward()` | correct |
| `diagonal_scatter(x, src, 0).add(1).sum().backward()` | **all zeros (BUG)** |
| `diagonal_scatter(x, src, 0).clone().sum().backward()` | **all zeros (BUG)** |

## Versions

- **PyTorch**: 2.13.0.dev20260425+cu126
- **Triton**: 3.2.0+git4b3bb1e8
- **CUDA**: 12.6
- **GPU**: Tesla T4 (sm_75)
- **Python**: 3.11

cc @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @aakhundov @coconutruben @jataylo

**Comment by irshadcc** [comment]
It seems that it is a issue with the decomposition. [Decomposition for diagonal_scatter](https://github.com/pytorch/pytorch/blob/5f0d7071a9bed278851b67b247ff7364a77ad9d0/torch/_refs/__init__.py#L4642)

> From Claude: 
> The bad path is reached in the backward of diagonal_scatter(x, src).sum(): AOT autograd lowers it to diagonal_scatter(expand(tangent, [N,N]), zeros). The expanded tangent has stride (0,0). The decomposition produced an output meta of f32[N,N][0,0], so Inductor allocated buf0 = empty_strided_cuda((1,1), (0,0)). Then the kernel that zeros the diagonal wrote 0.0 to that single aliased cell, and x.grad = reinterpret_tensor(buf0, (N,N), (0,0)) read back as all zeros.
>
> Eager dodged this because aten::diagonal_scatter is a primitive w.r.t. autograd in C++ and the C++ clone_preserve_strides correctly falls back to clone() for overlapping input.

I have a [fix](https://github.com/irshadcc/pytorch/commit/f51efbfabe7b5871476e53223ddb438febbc47cd) and would like to create a pull request if its okay

**Comment by jansel** [comment]
@irshadcc go ahead and submit a PR

**Comment by karthickai** [comment]
**🤖 PT2 Repro Agent**

I tried to reproduce this issue on PyTorch nightly (`2.13.0.dev20260514+cu130`) on NVIDIA H100 and confirmed it reproduces (3/3 runs failed).

All 3 runs reproduced the bug. Compiled x.grad is all zeros while eager x.grad is 1 everywhere except the diagonal; max_diff(x.grad)=1.0 across all runs. src.grad matches eager (max_diff=0.0) and the forward sum is correct, so only the backward gradient for x is wrong. The emulate_precision_casts=True mitigation did not fix the bug.

The compiler bisector narrowed the bug to the **inductor** backend; no single subsystem fully accounts for it (likely a multi-source issue accumulating across passes).

<details>
<summary>Error output</summary>

```
=== eager x.grad ===
tensor([[0., 1., 1., 1., 1., 1., 1., 1.],
        [1., 0., 1., 1., 1., 1., 1., 1.],
        [1., 1., 0., 1., 1., 1., 1., 1.],
        [1., 1., 1., 0., 1., 1., 1., 1.],
        [1., 1., 1., 1., 0., 1., 1., 1.],
        [1., 1., 1., 1., 1., 0., 1., 1.],
        [1., 1., 1., 1., 1., 1., 0., 1.],
        [1., 1., 1., 1., 1., 1., 1., 0.]], device='cuda:0')
=== compiled x.grad ===
tensor([[0., 0., 0., 0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0., 0., 0., 0.]], device='cuda:0')
=== eager src.grad ===
tensor([1., 1., 1., 1., 1., 1., 1., 1.], device='cuda:0')
=== compiled src.grad ===
tensor([1., 1., 1., 1., 1., 1., 1., 1.], device='cuda:0')
max_diff x.grad: 1.0
max_diff src.grad: 0.0
compiled x.grad all zero? True

/home/karthickps/.conda/envs/pytorch-nightly/lib/python3.13/site-packages/torch/_inductor/lowering.py:2231: FutureWarning: `torch._prims_common.check` is deprecated and will be removed in the future. Please use `torch._check*` functions instead.
  check(
/home/karthickps/.conda/envs/pytorch-nightly/lib/python3.13/site-packages/torch/_inductor/lowering.py:2231: FutureWarning: `torch._prims_common.check` is deprecated and will be removed in the future. Please use `torch._check*` functions instead.
  check(

ERROR conda.cli.main_run:execute(49): `conda run python repro.py` failed. (See above for error)

```
</details>

<details>
<summary>Repro script I used</summary>

```python
import os
import torch


def test_fn() -> bool:
    """Return True if NO bug (eager == compiled). False if bug present."""
    torch._dynamo.reset()
    backend = os.environ.get("TORCH_COMPILE_BACKEND", "inductor")

    device = "cuda" if torch.cuda.is_available() else "cpu"
    torch.manual_seed(42)

    N = 8

    def fn(x, src):
        return torch.diagonal_scatter(x, src, 0).sum()

    # Eager
    x1 = torch.randn(N, N, device=device, requires_grad=True)
    s1 = torch.randn(N, device=device, requires_grad=True)
    out_eager = fn(x1, s1)
    out_eager.backward()
    eager_x_grad = x1.grad.clone()
    eager_src_grad = s1.grad.clone()

    # Compiled
    x2 = x1.detach().clone().requires_grad_(True)
    s2 = s1.detach().clone().requires_grad_(True)
    out_compiled = torch.compile(fn, backend=backend)(x2, s2)
    out_compiled.backward()
    compiled_x_grad = x2.grad.clone()
    compiled_src_grad = s2.grad.clone()

    # Compare
    x_grad_ok = torch.allclose(eager_x_grad, compiled_x_grad, atol=1e-5)
    src_grad_ok = torch.allclose(eager_src_grad, compiled_src_grad, atol=1e-5)
    fwd_ok = torch.allclose(out_eager.detach(), out_compiled.detach(), atol=1e-5)

    return bool(x_grad_ok and src_grad_ok and fwd_ok)


if __name__ == "__main__":
    device = "cuda" if torch.cuda.is_available() else "cpu"
    torch.manual_seed(42)
    N = 8

    def fn(x, src):
        return torch.diagonal_scatter(x, src, 0).sum()

    x1 = torch.randn(N, N, device=device, requires_grad=True)
    s1 = torch.randn(N, device=device, requires_grad=True)
    fn(x1, s1).backward()

    torch._dynamo.reset()
    x2 = x1.detach().clone().requires_grad_(True)
    s2 = s1.detach().clone().requires_grad_(True)
    torch.compile(fn)(x2, s2).backward()

    print("=== eager x.grad ===")
    print(x1.grad)
    print("=== compiled x.grad ===")
    print(x2.grad)
    print("=== eager src.grad ===")
    print(s1.grad)
    print("=== compiled src.grad ===")
    print(s2.grad)

    max_diff_x = (x1.grad - x2.grad).abs().max().item()
    max_diff_s = (s1.grad - s2.grad).abs().max().item()
    print(f"max_diff x.grad: {max_diff_x}")
    print(f"max_diff src.grad: {max_diff_s}")
    print(f"compiled x.grad all zero? {bool((x2.grad == 0).all().item())}")

    try:
        ok = test_fn()
    except Exception:
        import traceback
        traceback.print_exc()
        os._exit(2)
    os._exit(0 if ok else 1)

```
</details>

<details>
<summary>My environment</summary>

```
Collecting environment information...
PyTorch version: 2.13.0.dev20260514+cu130
Is debug build: False
CUDA used to build PyTorch: 13.0
ROCM used to build PyTorch: N/A

OS: CentOS Stream 9 (x86_64)
GCC version: (GCC) 11.5.0 20240719 (Red Hat 11.5.0-14)
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.34

Python version: 3.13.12 | packaged by Anaconda, Inc. | (main, Feb 24 2026, 16:13:31) [GCC 14.3.0] (64-bit runtime)
Python platform: Linux-6.16.1-0_fbk2_hardened_0_gf40efc324cc8-x86_64-with-glibc2.34
Is CUDA available: True
CUDA runtime version: 12.8.93
CUDA_MODULE_LOADING set to: 
GPU models and configuration: GPU 0: NVIDIA H100
Nvidia driver version: 580.126.09
cuDNN version: Could not collect
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True
Caching allocator config: N/A

CPU:
Architecture:                            x86_64
CPU op-mode(s):                          32-bit, 64-bit
Address sizes:                           52 bits physical, 57 bits virtual
Byte Order:                              Little Endian
CPU(s):                                  46
On-line CPU(s) list:                     0-45
Vendor ID:                               AuthenticAMD
Model name:                              AMD EPYC 9654 96-Core Processor
CPU family:                              25
Model:                                   17
Thread(s) per core:                      1
Core(s) per socket:                      46
Socket(s):                               1
Stepping:                                1
BogoMIPS:                                4792.79
Flags:                                   fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm rep_good nopl cpuid extd_apicid tsc_known_freq pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand hypervisor lahf_lm cmp_legacy svm cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw perfctr_core ssbd ibrs ibpb stibp ibrs_enhanced vmmcall fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves avx512_bf16 clzero xsaveerptr wbnoinvd arat npt lbrv nrip_save tsc_scale vmcb_clean flushbyasid pausefilter pfthreshold v_vmsave_vmload vgif vnmi avx512vbmi umip pku ospke avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg avx512_vpopcntdq la57 rdpid fsrm flush_l1d arch_capabilities
Virtualization:                          AMD-V
Hypervisor vendor:                       KVM
Virtualization type:                     full
L1d cache:                               2.9 MiB (46 instances)
L1i cache:                               2.9 MiB (46 instances)
L2 cache:                                23 MiB (46 instances)
L3 cache:                                736 MiB (46 instances)
NUMA node(s):                            1
NUMA node0 CPU(s):                       0-45
Vulnerability Gather data sampling:      Not affected
Vulnerability Ghostwrite:                Not affected
Vulnerability Indirect target selection: Not affected
Vulnerability Itlb multihit:             Not affected
Vulnerability L1tf:                      Not affected
Vulnerability Mds:                       Not affected
Vulnerability Meltdown:                  Not affected
Vulnerability Mmio stale data:           Not affected
Vulnerability Old microcode:             Not affected
Vulnerability Reg file data sampling:    Not affected
Vulnerability Retbleed:                  Not affected
Vulnerability Spec rstack overflow:      Vulnerable
Vulnerability Spec store bypass:         Vulnerable
Vulnerability Spectre v1:                Vulnerable: __user pointer sanitization and usercopy barriers only; no swapgs barriers
Vulnerability Spectre v2:                Vulnerable; IBPB: disabled; STIBP: disabled; PBRSB-eIBRS: Not affected; BHI: Not affected
Vulnerability Srbds:                     Not affected
Vulnerability Tsa:                       Vulnerable
Vulnerability Tsx async abort:           Not affected

Versions of relevant libraries:
[pip3] mypy_extensions==1.1.0
[pip3] numpy==2.4.4
[pip3] nvidia-cublas==13.1.1.3
[pip3] nvidia-cuda-cupti==13.0.85
[pip3] nvidia-cuda-nvrtc==13.0.88
[pip3] nvidia-cuda-runtime==13.0.96
[pip3] nvidia-cudnn-cu13==9.20.0.48
[pip3] nvidia-cufft==12.0.0.61
[pip3] nvidia-curand==10.4.0.35
[pip3] nvidia-cusolver==12.0.4.66
[pip3] nvidia-cusparse==12.6.3.3
[pip3] nvidia-cusparselt-cu13==0.8.1
[pip3] nvidia-nccl-cu13==2.29.7
[pip3] nvidia-nvjitlink==13.0.88
[pip3] nvidia-nvtx==13.0.85
[pip3] torch==2.13.0.dev20260514+cu130
[pip3] torchao==0.17.0
[pip3] torchaudio==2.11.0.dev20260515+cu130
[pip3] torchdata==0.11.0
[pip3] torchmetrics==1.0.3
[pip3] torchrec==1.7.0.dev20260413+cu130
[pip3] torchtune==0.6.1
[pip3] torchvision==0.28.0.dev20260515+cu130
[pip3] triton==3.7.0+git88b227e2
[conda] Could not collect
```
</details>

---
If the repro agent produced an incorrect result (wrong classification, bad repro script, missed the actual bug, etc.), add the label `repro:agent:h100:incorrect` to the issue.
